src/buf/fixed/handle.rs

```diff
@@ -44,59 +42,51 @@ impl Drop for FixedBuf {
         // maintained accordingly to the safety contracts on
         // Self::new and IoBufMut.
         unsafe {
-            registry.check_in(self.index, self.buf.len());
+            registry.check_in(self.buf.index, self.buf.init_len);
         }
     }
 }
 
 impl FixedBuf {
     // Safety: Validity constraints must apply to CheckedOutBuf members:
-    // - iovec must refer to an array allocated by Vec<u8>;
     // - the array will not be deallocated until the buffer is checked in;
     // - the data in the array must be initialized up to the number of bytes
     //   given in init_len.
     pub(super) unsafe fn new(registry: Rc<RefCell<dyn FixedBuffers>>, data: CheckedOutBuf) -> Self {
-        let CheckedOutBuf {
-            iovec,
-            init_len,
-            index,
-        } = data;
-        let buf = Vec::from_raw_parts(iovec.iov_base as _, init_len, iovec.iov_len);
         FixedBuf {
             registry,
-            buf: ManuallyDrop::new(buf),
-            index,
+            buf: data,
         }
     }
 
     /// Index of the underlying registry buffer
     pub fn buf_index(&self) -> u16 {
-        self.index
+        self.buf.index
     }
 }
 
 unsafe impl IoBuf for FixedBuf {
     fn stable_ptr(&self) -> *const u8 {
-        self.buf.as_ptr()
+        self.buf.iovec.iov_base as _
     }
 
     fn bytes_init(&self) -> usize {
-        self.buf.len()
+        self.buf.init_len
     }
 
     fn bytes_total(&self) -> usize {
-        self.buf.capacity()
+        self.buf.iovec.iov_len
     }
 }
 
 unsafe impl IoBufMut for FixedBuf {
     fn stable_mut_ptr(&mut self) -> *mut u8 {
-        self.buf.as_mut_ptr()
+        self.buf.iovec.iov_base as _
     }
 
     unsafe fn set_init(&mut self, pos: usize) {
-        if self.buf.len() < pos {
-            self.buf.set_len(pos)
+        if self.buf.iovec.iov_len >= pos {
```

Comment:
I found this interesting @mzabaluev . Is the original in error? You can't ever decrease the number of initialized bytes?
Changed to what I believe to be the intention, which is you cannot have more initialized bytes than the capacity.

Reply:
No, `init_len` is for tracking the length of the initialized bytes, and once they are known to be initialized, there is no reason to declare them uninitialized. The original code is correct (for `Vec`).

Reply:
Whilst I guess thats true, it goes against what I would expect

```rust
file1.read_at(buf,0);
file2.read_at(buf,0);
```
I guess i'd expect `bytes_init` to represent the number of bytes from file 2 at this point.